### PR TITLE
Email for subset of refs -- with careful handling of the logic for detection of 'new' commits

### PR DESCRIPTION
**Pull request by newren**
There are many reasons folks might want to only have emails sent out for a subset of the refs in the repository:

1) A common workflow for folks is to have code reviews occur by pushing to a special branch (e.g. refs/heads/feature/X) or reference (refs/changes/NN/NNNNN/N).  The code reviews are then handled via webby-tools (Atlassian Stash, Gerrit, etc.), not via email.  Folks in general will already be pinged for the relevant code reviews, and they don't want extra emails about them, nor do they want emails about unofficial changes from other folks.  So, providing a way to have git-multimail only send emails about a relevant subset of the repository is useful.

2) Another common workflow for people working on not-official-stuff is to push to locations such as refs/sandbox/$USER/$BRANCH.  While people could pull directly from each other, many seem to like the workflow of pushing to a special place in the central/official repository.  However, as this stuff is considered personal sandbox stuff, it makes no sense to broadcast emails about it.

3) Even if it weren't for these special workflows, the refs/notes/ hierarchy are somewhat of an implementational detail which users never look at directly.  Emailing about updates to these refs is simply confusing.

4) Finally, various git servers may have implementation reasons for storing other special refs (e.g. refs/cache-automerge/* in Gerrit) under various circumstances, and these changes should not be sent to users either.

Because of all these reasons, it is very valuable to provide ways for users to mark a subset of refs as relevant, via either an exclusion or inclusion regex.  The thing to be careful about is that emails for revisions do go out once those revisions become included in the 'official' part of the repository.

(I originally developed this in a couple stages in order to generalize it more as more was needed.  If squashing some of these commits together makes more sense to you, just let me know.)

### CODE REVIEW

**Comment by moy** [comment]
No time for a detailed review, but how does this compare to #75 and #15 ? It seems to me that they are trying to solve the same issue. Probably this one is the most complete. Please read the discussion in these PR and see if you address all the issues raised there.

Thanks,

File: git-multimail/README [hunk 1]
```diff
@@ -388,6 +394,18 @@ multimailhook.replyToRefchange
...
+multimailhook.refFilterExclusionRegex
+
+    Regular expressions that can be used to limit refs for which email
+    updates will be sent.  It is an error to specify both an inclusion
+    and an exclusion regex.  If an inclusion regex is specified, emails
+    will only be sent for refs which match this regex.  If an exclusion
+    regex is specified, emails will be sent for all refs except those
+    that match this regex (or that match a predefined regex specific
+    to the environment, such as "^refs/notes" for most environments and
+    "^refs/notes|^refs/changes" for the gerrit environment).
```
Comment (moy) [inline git-multimail/README]: Can we change this to "whitespace-separated list of regex"? The `|` works, but makes the expression less readable (it might get cryptic if we have several non-trivial regex separated by `|`).

Note that we don't need to do that now, since we can add the whitespace-separated list idea later without breaking compatibility.

File: git-multimail/README [hunk 2]
```diff
@@ -388,6 +394,18 @@ multimailhook.replyToRefchange
     - The value "none", in which case the Reply-To: field will be
       omitted.
 
+multimailhook.refFilterInclusionRegex
+multimailhook.refFilterExclusionRegex
+
+    Regular expressions that can be used to limit refs for which email
+    updates will be sent.  It is an error to specify both an inclusion
+    and an exclusion regex.  If an inclusion regex is specified, emails
```
Comment (moy) [inline git-multimail/README]: I'm wondering whether we should accept both options to be set, possibly with a third option to specify the order (like Apache's `Order Deny,Allow`). Or maybe this would be overkill.

File: git-multimail/git_multimail.py [hunk 3]
```diff
@@ -2373,18 +2439,34 @@ def send_emails(self, mailer, body_filter=None):
...
 
 
+def include_ref(refname, ref_filter_regex, is_inclusion_filter):
+  return not (bool(ref_filter_regex.search(refname)) ^ is_inclusion_filter)
+  ## More readable version:
+  #does_match = bool(re.search(ref_filter_regex))
+  #if is_inclusion_filter:
+  #  return does_match
+  #else:  # exclusion filter -- we include the ref if the regex doesn't match
+  #  return not does_match
```
Comment (moy) [inline git-multimail/git_multimail.py]: I do prefer the "more readable" one ;-).

**Comment by moy** [comment]
In my version (#15), I used glob instead of regexp. I tend to prefer glob, which are simpler to use (I suspect we'll use only `.*` and `|`, and if so a list of glob just works).

Any opinion on that?

**Comment by moy** [comment]
I had forgotten about this, but re-reading the comments for #15, I remember that @mhagger had a really good idea to describe the configuration using subsections: https://github.com/git-multimail/git-multimail/pull/15#issuecomment-14676468

I'm not sure what is the best option now:

1. merge the PR more or less as-is, and think about the subsection idea later (or never)

   **pro:** we get the new feature soon, with little effort on our side.
   **con:** if we add the subsection idea, we'll have two ways to say the same thing in the config file => longer documentation, more code to maintain, ...

2. refuse merging this PR until we have the subsection implemented. The issue is that we may never get there.

**Comment by moy** [comment]
> The thing to be careful about is that emails for revisions do go out once those revisions become included in the 'official' part of the repository.

This is actually debatable. In the use-case which lead to #15 for me, I wanted to have a long-lived branch and not disturb other people when working on it. I didn't want to patch-bomb my co-workers with 100+ emails at the time of merging. Another example in the discussion was debian packages, where the upstream branch should be ignored completely.

So, to me, this should be configurable. If we stick to the current way of configuring, this could be a `multimailhook.refFilterMode` with values `permanent` (never send the change email) or `untilMerge` (send the email when the commit reaches a non-ignored ref).

Any opinion?

File: git-multimail/README (hunk 1 above)
Comment (mhagger) [inline git-multimail/README]: Generally I think that allowing this setting to be multivalued would be cleaner than allowing space-separated values that are pulled apart and glued together within the code:

    [multimailhook]
        refFilterExclusionRegex = ^refs/notes
        refFilterExclusionRegex = ^refs/changes

Either way you would probably want to glue the regexps together internally into a single regexp using something like

    '|'.join(
        '(?:%s)' % (fragment,)
        for fragment in fragments
        )

Comment (moy) [inline git-multimail/README]: > allowing this setting to be multivalued would be cleaner than allowing space-separated values

You don't have to chose actually: it's rather easy to accept both.

**Comment by mhagger** [comment]
Re glob vs. regexp:

* globs are easier to use, but not quite as flexible. What is the most complicated use case people can think of? Probably something like "match release tags but not other tags".
* globs are closer to Git refspecs, which also appear in the configuration file. Though if the variable name includes the string "regex" I doubt that people will be confused.
* Please note that Python globs are meant for single pathname components and don't treat `/` specially. This is not ideal for matching reference names. For example, the pattern 'a/*/b' would match 'a/x/y/z/b'. 
* Git's glob patterns for filenames (e.g., for `.gitattributes` files) are much nicer and more flexible, but I don't think there is an interface to use them outside of git code.

Of course we could implement our own glob-matching code, for example translating from whatever kind of glob syntax we want to allow into regexps internally and using those.

Bottom line: I have no strong opinion either way. I think using `fnmatch.fnmatchcase()` directly wouldn't be ideal because of the limitations mentioned above; regexps would be preferable. But either regexps or sane glob patterns, I think, would be totally fine.

**Comment by moy** [comment]
I didn't know about Python's globs specifics. Well, then I'm starting to get convinced that regexp are the way to go.

File: git-multimail/README (hunk 1 above)
Comment (newren) [inline git-multimail/README]: Actually, neither of these variables is a set/list/whatever of regexes.  It's a single regex.  But some folks only know very basic regexes (e.g. only being aware of the three special characters: ., +, and *), and I wanted to describe this variable while simultaneously hinting how they could match multiple things -- without being overly verbose.  I figured those who knew better would just say "A '|'-separated list of regexes for matching any one of the regexes?  How cute; that's just the normal regex syntax for matching multiple different things, so it's really just one regex."

Clearly, my wording didn't have the intended effect.  However, I like the fact that it's just a regex. If we make it a space separated list or multiple config values, then it loses that quality.  If that's the way you want it, I'll implement it that way, but I personally much prefer that each of these variables is just a plain python regex.

File: git-multimail/README (hunk 2 above)
Comment (newren) [inline git-multimail/README]: Actually, I think that sounds like a really cool idea.

File: git-multimail/git_multimail.py (hunk 3 above)
Comment (newren) [inline git-multimail/git_multimail.py]: Yeah, I tend to prefer 1 line of code over 5, but I suspected this one was pushing the limits.  I'll be happy to switch it.

**Comment by newren** [comment]
Is there a way to respond to individual comments in github PRs that weren't specific to a certain file/line?  I've got a bunch to respond to, hopefully context isn't lost:

In regards to the question about adding a multimailhook.refFilterMode with values of permanent or untilMerge; your usecase description makes sense to me, and I can see why you want a different behavior than I do.  I think adding the flexibility to cover both usecases makes sense.  (Also, if we decide to implement this behavior via subsections, the implemention may need to be something closer to rhertzog's suggestion in #15 of a 'notification' setting per subsection with possible values of 'none', 'none-except-on-merge', or 'all'.)

In regards to globs vs. regexes: It sounds like others are already leaning towards regexes now, which makes me happy.  I prefer them for this particular feature, and now for even more reasons.  :-)

In regards to the idea of subsections: That actually does sound like a pretty cool idea, and perhaps makes the inclusion|exclusion ordering even more versatile than something like Apache's "Order Allow, Deny".  Would it apply just to whether to send emails, or would it start encompassing all variables (e.g. mailingList, announceShortLog, mailer, administrator, emailPrefix, maxCommitEmails)?  I'm curious whether any of those (mailer?) are particularly hard to handle with such a scheme.  Also, are there valid regexes which would not be allowed as the name of a subsection?  Anyway, in response to the question about whether we should block this feature until we have it implemented via subsections, I agree that's a good tough question.  Let me take a few days to think about that (and maybe see if I can convince my boss to let me work on it; there's at least a chance even if it's not as big as I'd like).

In regards to comparing to #15 and #75: Yeah, I agree that this one is the most feature complete.  I think it has everything #75 does, and almost everything #15 does -- it'd just need the refFilterMode to toggle between our different behaviors after merges.  Other than that, I think the only big question left at this point is whether to re-implement via sub-sections.  Is there anything important I missed?

**Comment by moy** [comment]
I don't know any way to do proper threading in GitHub, so let's mix the topics ;-).

About subsections: Git is already using them to store URL, and refnames are less problematic than URL wrt special characters & such, so it should just work. We would have to be careful because just calling `git config -l` may yield an ambiguous output when both the value and the subsection can contain the characters `=` and `.`, like:

    [foo "bar=boz"]                                                                                                                                                
         x = .bar=boz
    [foo "bar=boz.x="]                                                                                                                                             
         bar=boz

    $ git config -l | grep boz                                                                                                                           
    foo.bar=boz.x=.bar=boz
    foo.bar=boz.x=.bar=boz

File: git-multimail/README (hunk 1 above)
Comment (moy) [inline git-multimail/README]: newren <notifications@github.com> writes:

> Actually, neither of these variables is a set/list/whatever of
> regexes.

Currently, yes. My point is that it would be easier to use if it became
a list of regexp. Just like `grep` can accept multiple `-e` options so
you can write

    grep -e foo -e bar

instead of

    grep -e 'foo\|bar'

I find

    [multimailhook]
        refFilterExclusionRegex = ^refs/notes
        refFilterExclusionRegex = ^refs/changes

easier to read than

    [multimailhook]
        refFilterExclusionRegex = ^refs/notes\|^refs/changes

plus I don't have to remember which kind of regurar expression, what
kind of quoting I should use and all in the first case (| Vs \| Vs \\|
depending on the context).

And I could give more complex examples like

    [multimailhook]
        refFilterExclusionRegex = ^refs/heads/\(foo\|bar\)/boz\|^refs/changes

which get really hard to read when hand-assembled into a single regular
expression, even though the expressivity is the same.

Now, the good news is that both "space separated list of regexp" and
"multi-valued config variable" can be added later as an extension to
your version. The multi-valued config variable system would break the
compatibility only for users relying on the "last one wins" semantics.

> Clearly, my wording didn't have the intended effect. However, I like
> the fact that it's just a regex. If we make it a space separated list
> or multiple config values, then it loses that quality.

Sorry, but I see no real argument here. If you like writting single
regexp, you'll still be able to do so.

-- 
Matthieu Moy
http://www-verimag.imag.fr/~moy/

**Comment by moy** [comment]
> In regards to the idea of subsections: [...] Let me take a few days to think about that

Any new thoughts on the topic? I'll try to tag a 1.1 release in June, so if you wish this topic to be in it, it should move on soon. It's not a problem if we don't merge anything before the release, it can also wait the next one. Excluding some refs is one of the most asked for feature it seems, so it could deserve a release on its own.

**Comment by moy** [comment]
Actually, about the "glob" Vs "regex", I think the best option would be to have a pattern that can be passed as argument to `git for-each-ref`, which is essentially a glob with some additional magic:

    $ git for-each-ref refs/heads/
    32ba5b7e503c7bc7eaea233d67a649fc4da81298 commit refs/heads/1.1
    129dcd3150681ef221de610eedffb29376911f48 commit refs/heads/master
    
    $ git for-each-ref 'refs/remotes/*/master' refs/tags
    5e432a55fa2cccc161cac7583e552c3b0451e783 commit refs/remotes/moy/master
    129dcd3150681ef221de610eedffb29376911f48 commit refs/remotes/origin/master
    1b32653bafc4f902535b9fc1cd9cae911325b870 commit refs/tags/1.0.0
    8c3aaafa873bf10de8dddf1d202c449b3eff3b42 commit refs/tags/1.0.2
    ...

It's much simpler to write than regexp (no need for ^-anchoring, a prefix-match is straightforward and more complex matches are possible with `*`. I can't imagine a real-life case where this would not be sufficient), easy to implement (by calling `git for-each-ref`), and it makes it easy to test the pattern before writing them to the config file (just run `git for-each-ref` by hand and see which branch match).

One drawback of the approach though is that calling `git for-each-ref` is O(n), it may be slow in repositories with really lots of refs, but a quick test shows that even with 10k-refs, we can still do this in <0.2 sec (even less with packed refs).

**Comment by moy** [comment]
I created a new PR (#100) for this.
